### DISCUSSION TITLE
Next.js App with @supabase/auth-helpers-nextjs middleware

### DISCUSSION THREAD

**Discussion by JustinBrenner** [supabase/supabase#13879] [Questions] [answered]
Hello All, 

TIA for assistance on this. I am somewhat a intermediate with supabase but wanted to try out some new features with the auth-helpers beta and middleware in Nextjs. My use case from what i thought was pretty simple:

I have a Nextjs 13 app with a /login and /register page and then i have a /app directory that has a layout tied to each page in that directory but whenever i sign in using the /login route i have a session returned to the client but it is like middleware does not want to redirect to the /app directory with the authenticated user. Im not sure what i am doing wrong but figured that cookies may help but have no idea where to start with that. Can y'all please help me? 

 @dijonmusters @thorwebdev 
 
 **Middleware code**
 
`import { NextRequest, NextResponse } from "next/server";
import { createMiddlewareSupabaseClient } from "@supabase/auth-helpers-nextjs";

export async function middleware(req: NextRequest) {
    console.log('middleware, no session')
    const adminPath = "/app";
    const apiAdminPath = "/api/app";
  
    const res = NextResponse.next();
    const supabase = createMiddlewareSupabaseClient({ req, res });
    const {
      data: { session },
    } = await supabase.auth.getSession();
  console.log()
    if (!session ) { //|| session?.user.user_metadata?.role !== "admin"
        console.log('middleware, no session')
      if (req.nextUrl.pathname.startsWith(apiAdminPath)) {
        return new NextResponse(
          JSON.stringify({ message: "authorization failed" }),
          { status: 403, headers: { "Content-Type": "application/json" } }
        );
      } else if (req.nextUrl.pathname.startsWith(adminPath)) {
        const redirectUrl = req.nextUrl.clone();
        redirectUrl.pathname = "/login";
        return NextResponse.redirect(redirectUrl);
      }
    }
    console.log('middleware, session exists')
  }
  
  export const config = {
    matcher: ["/api/app/:path*", "/app/:path*"],
  };`

**Comment by dijonmusters** [comment, marked as answer]
Hey! This looks good to me and should be working. A couple of things I would check:

1. Make sure you are using `createBrowserSupabaseClient` when you are logging the user in
2. Make sure you are using `createServerComponentSupabaseClient` to access your session in a Server Component
3. Remove that first `console.log` in your middleware file, as it makes it seem like you never have a session

I have put together a repo using this middleware function that can hopefully steer you in the right direction 👍

https://github.com/dijonmusters/discussion-13879

Also, would recommend this series of videos to come up to speed with all the new Next.js App Directory stuff 🙂 

https://www.youtube.com/watch?v=Bh1TOpOcGJQ&list=PL5S4mPUpp4OuMkz8qqJpiTSaX9Fp0429T&index=3

Let me know if you have any other issues 👋

**Reply by roker15** [reply]
It never works with google signIn, can you give me an example for auth sign in?
```
const handleGoogleLogin = async () => {
    const { error } = await supabase.auth.signInWithOAuth({
      provider: "google",
      options: { redirectTo: `http://www.localhost:3000/redirect` },
    });

    if (error) {
      console.log({ error });
    }
    router.refresh();
  };
```
`router.refresh` never triggers, and i need to refresh manually.

My middleware.ts
```
export async function middleware(req: NextRequest) {

  const res = NextResponse.next();
  const supabase = createMiddlewareClient<Database>({ req, res });
  const {
    data: { session },
  } = await supabase.auth.getSession();
  return res;
}
```

**Comment by derekjwilliams** [comment]
If you ended up on this page please be aware that these functions have been deprecated, see https://ld.reddit.com/r/Supabase/comments/17j0utu/nextjsauthhelpers_already_deprecated/k72h58i/ for a description
